Add Site property to xADDomainController. Close #76

gordonbondon opened the pull request:
- Add possibility to set Site for domain controller
- Add tests for new Site functionality
- Updates README
- Closes #76

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/powershell/xactivedirectory/109)
<!-- Reviewable:end -->

msftclas commented:
Hi __@gordonbondon__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

narrieta commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[DSCResources/MSFT_xADDomainController/MSFT_xADDomainController.psm1, line 117 \[r1\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KOBtO8XefWLhfz0xxjE:-KOBtO8XefWLhfz0xxjF:-501343635) ([raw file](https://github.com/powershell/xactivedirectory/blob/f8cdd0c50652ccc3fcc3bbd3e22f617fc9f26015/DSCResources/MSFT_xADDomainController/MSFT_xADDomainController.psm1#L117)):*
> ```PowerShell
>             {
>                 $site = Get-ADReplicationSite -Identity $SiteName -Server $existingDC.HostName -Credential $DomainAdministratorCredential
>             }
> ```

$site is never used

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 59 \[r1\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KOBvflsLWvedPlJcFTy:-KOBvflsLWvedPlJcFTz:-1430084021) ([raw file](https://github.com/powershell/xactivedirectory/blob/f8cdd0c50652ccc3fcc3bbd3e22f617fc9f26015/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L59)):*
> ```PowerShell
> 
>                 #function Get-ADDomain {[CmdletBinding()] param($Identity, $Credential)}
>                 #function Get-ADDomainController {[CmdletBinding()] param($Identity, $Credential)}
> ```

These 2 lines seem to be leftovers

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 155 \[r1\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KOBwck2NBn5SiRYZFM0:-KOBwck2NBn5SiRYZFM1:-1720049689) ([raw file](https://github.com/powershell/xactivedirectory/blob/f8cdd0c50652ccc3fcc3bbd3e22f617fc9f26015/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L155)):*
> ```PowerShell
>                 function Move-ADDirectoryServer {
>                     param (
>                         $Idintity, $Site, $AuthType, $Confirm, $Credential, $Server
> ```

"Identity"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

narrieta commented:
I added only 2 minor comments. Other that that it looks good.

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-:-KOBwmj6R6pl9nqcDLKe:-1059986569)*
<!-- Sent from Reviewable.io -->

gordonbondon commented:
Newbie question. Should i squash fixup commits or it will be done later when merging by someone from DSC team?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-:-KOEldN6Sw8l7VeCb2lS:1000146763)*
<!-- Sent from Reviewable.io -->

narrieta commented:
@gordonbondon : could you squash them?

gordonbondon commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[DSCResources/MSFT_xADDomainController/MSFT_xADDomainController.psm1, line 117 \[r1\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KOBtO8XefWLhfz0xxjE:-KOHYhqbj1sSwLPT2vHP:-1861261135) ([raw file](https://github.com/powershell/xactivedirectory/blob/f8cdd0c50652ccc3fcc3bbd3e22f617fc9f26015/DSCResources/MSFT_xADDomainController/MSFT_xADDomainController.psm1#L117)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> $site is never used
</details>
Probably it will make more sense if I move Site check to a separate private function?

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 59 \[r1\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KOBvflsLWvedPlJcFTy:-KOHYWF4iK8i0_-XDZv-:-499146009) ([raw file](https://github.com/powershell/xactivedirectory/blob/f8cdd0c50652ccc3fcc3bbd3e22f617fc9f26015/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L59)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> These 2 lines seem to be leftovers
</details>
Done.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 155 \[r1\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KOBwck2NBn5SiRYZFM0:-KOHYVANUKlghoyq-Okk:-499146009) ([raw file](https://github.com/powershell/xactivedirectory/blob/f8cdd0c50652ccc3fcc3bbd3e22f617fc9f26015/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L155)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> "Identity"
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

gordonbondon commented:
@narrieta yes, after we've done the review i can squash it all to one commit.

narrieta commented:
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[DSCResources/MSFT_xADDomainController/MSFT_xADDomainController.psm1, line 117 \[r1\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KOBtO8XefWLhfz0xxjE:-KOHs8nhjKAD37NiVjuj:-2121497265) ([raw file](https://github.com/powershell/xactivedirectory/blob/f8cdd0c50652ccc3fcc3bbd3e22f617fc9f26015/DSCResources/MSFT_xADDomainController/MSFT_xADDomainController.psm1#L117)):*
<details><summary><i>Previously, gordonbondon (Artem Iarmoliuk) wrote…</i></summary>
> Probably it will make more sense if I move Site check to a separate private function?
</details>
I only meant that we are assigning to $site, but I think the variable is never used. Maybe remove the assignment? Private function OK as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

narrieta commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[DSCResources/MSFT_xADCommon/MSFT_xADCommon.ps1, line 585 \[r4\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KOLPFvahPHUtKZBgixt:-KOLPFvahPHUtKZBgixu:1893894274) ([raw file](https://github.com/powershell/xactivedirectory/blob/12d4c611f62c66996a92190b37e0be09cf8c1db7/DSCResources/MSFT_xADCommon/MSFT_xADCommon.ps1#L585)):*
> ```PowerShell
>         [Parameter()]
>         [System.Management.Automation.PSCredential]
>         [System.Management.Automation.CredentialAttribute()]
> ```

Let's not use the Credential attribute, since it can trigger a prompt and DSC does not support interactive input.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

gordonbondon commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[DSCResources/MSFT_xADCommon/MSFT_xADCommon.ps1, line 585 \[r4\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KOLPFvahPHUtKZBgixt:-KOOwIzP3yuAe6Q7b7O3:b-896fix) ([raw file](https://github.com/powershell/xactivedirectory/blob/12d4c611f62c66996a92190b37e0be09cf8c1db7/DSCResources/MSFT_xADCommon/MSFT_xADCommon.ps1#L585)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> Let's not use the Credential attribute, since it can trigger a prompt and DSC does not support interactive input.
</details>
Done.

---

*[DSCResources/MSFT_xADDomainController/MSFT_xADDomainController.psm1, line 117 \[r1\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KOBtO8XefWLhfz0xxjE:-KOLZg_GLkVtS-fpXuAW:b-ytmdxk) ([raw file](https://github.com/powershell/xactivedirectory/blob/f8cdd0c50652ccc3fcc3bbd3e22f617fc9f26015/DSCResources/MSFT_xADDomainController/MSFT_xADDomainController.psm1#L117)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> I only meant that we are assigning to $site, but I think the variable is never used. Maybe remove the assignment? Private function OK as well.
</details>
Done.
I should have been more clear. I've used $site variable to check for Site existence. Now this check is moved to separate function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

gordonbondon commented:
Not sure why my tests are always hanging in Appveyor. They pass perfectly fine on Win 10 1511.2 and Win Server 2012 R2 with latest updates, in pester 3.4, 3.4.1 and 3.4.2. Any ideas what can cause this?

narrieta commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 29 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KOQyY40dxbcEtFv2XEx:-KOQyY40dxbcEtFv2XEy:b-j2tudh) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L29)):*
> InModuleScope

This was in our original test template, but we are moving away from wrapping the entire test suite within InModule. The template has already been updated. Could you remove from here and use to wrap any code that needs access to the module's private data/functions (if any)? Thanks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

narrieta commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 121 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KOR0dpMMSPnCetg_kxP:-KOR0dpNKzsi1gFMXxLw:b5ets3j) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L121)):*
> ```PowerShell
>                 Mock Get-ADDomainController { return $stubDomainController }
>                 Mock Test-ADReplicationSite { return $false }
>                 { Test-TargetResource @testDefaultParams -DomainName $correctDomainName -SiteName $correctSiteName } | Should Throw
> ```

If possible can we check for a specific exception?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

narrieta commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 136 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KORWbzPcHYpOLe6hjwe:-KORWbzPcHYpOLe6hjwf:b-px9vp0) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L136)):*
> Install-ADDSDomainController

Do we need this function? We are movking Install-ADDSDomainController a few lines below.

Move-ADDirectoryServer too

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

gordonbondon commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 29 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KOQyY40dxbcEtFv2XEx:-KOYXOyjCHgczCid2v4t:b-b7ft2f) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L29)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> > InModuleScope
> 
> This was in our original test template, but we are moving away from wrapping the entire test suite within InModule. The template has already been updated. Could you remove from here and use to wrap any code that needs access to the module's private data/functions (if any)? Thanks
</details>
Probably I am doing something wrong in my tests but when i remove InModuleScope all mocks are ignored and tests try to call original functions instead. Any advice on what should I do  to fix this? I was not able to find something in Pester docs.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 121 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KOR0dpMMSPnCetg_kxP:-KOYXecRutu_2flOm6P0:b-896fix) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L121)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> If possible can we check for a specific exception?
</details>
Done.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 136 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KORWbzPcHYpOLe6hjwe:-KOYXhtSEIkj9gsYMBGC:bs9vuj2) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L136)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> > Install-ADDSDomainController
> 
> Do we need this function? We are movking Install-ADDSDomainController a few lines below.
> 
> Move-ADDirectoryServer too
</details>
Install-ADDSDomainController cmdlet is only available on server. So we need to fake it to run tests on client machines.
Fake Move-ADDirectoryServer was removed. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

narrieta commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 29 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KOQyY40dxbcEtFv2XEx:-KOfCnRAOWYI-VIB8INI:b-5400br) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L29)):*
<details><summary><i>Previously, gordonbondon (Artem Iarmoliuk) wrote…</i></summary>
> Probably I am doing something wrong in my tests but when i remove InModuleScope all mocks are ignored and tests try to call original functions instead. Any advice on what should I do  to fix this? I was not able to find something in Pester docs.
</details>
You would need to add -ModuleName to the definition of the mock. There are some examples here: https://github.com/pester/Pester/wiki/Unit-Testing-within-Modules

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

narrieta commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 136 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KORWbzPcHYpOLe6hjwe:-KOfDHMPJApXUT4uUSEN:b8j727t) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L136)):*
<details><summary><i>Previously, gordonbondon (Artem Iarmoliuk) wrote…</i></summary>
> Install-ADDSDomainController cmdlet is only available on server. So we need to fake it to run tests on client machines.
> Fake Move-ADDirectoryServer was removed. 
</details>
What I mean is that we have a function definition

    function Install-ADDSDomainController { ... }

and a mock of the same function

   Mock Install-ADDSDomainController -MockWith {} -ParameterFilter { $SiteName -eq $correctSiteName }

The mock should be enough... maybe I am missing something?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

johlju commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 136 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KORWbzPcHYpOLe6hjwe:-KOfHnlUpnOX92kMQ468:b-4e24oc) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L136)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> What I mean is that we have a function definition
> 
>     function Install-ADDSDomainController { ... }
> 
> and a mock of the same function
> 
>    Mock Install-ADDSDomainController -MockWith {} -ParameterFilter { $SiteName -eq $correctSiteName }
> 
> The mock should be enough... maybe I am missing something?
> 
</details>
@narrieta Actually, I found when mocking, it has to be able to call the actual function to be able to mock. So if `Install-ADDSDomainController` exist in a module that is not part of  the resource, and the module does not exist on the machine your run the test on, then you have to fake the cmdlet. I do it my faking the entire module, like below, but adding it as a function worked too (I was afraid of scopes, so I mocked the entire module).
```
        Get-Module -Name MockAD | Remove-Module -Force
        New-Module -Name MockAD -ScriptBlock {
            function Install-ADDSDomainController { return }

            Export-ModuleMember -Function *-ADDS*
        } | Import-Module -Force

        Mock -CommandName Install-ADDSDomainController -MockWith {} -ModuleName $script:DSCResourceName -Verifiable
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

narrieta commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 136 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KORWbzPcHYpOLe6hjwe:-KOfe4W3RERRJH-xyrDS:b-nmroam) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L136)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> @narrieta Actually, I found when mocking, it has to be able to call the actual function to be able to mock. So if `Install-ADDSDomainController` exist in a module that is not part of  the resource, and the module does not exist on the machine your run the test on, then you have to fake the cmdlet. I do it my faking the entire module, like below, but adding it as a function worked too (I was afraid of scopes, so I mocked the entire module).
> ```
>         Get-Module -Name MockAD | Remove-Module -Force
>         New-Module -Name MockAD -ScriptBlock {
>             function Install-ADDSDomainController { return }
> 
>             Export-ModuleMember -Function *-ADDS*
>         } | Import-Module -Force
> 
>         Mock -CommandName Install-ADDSDomainController -MockWith {} -ModuleName $script:DSCResourceName -Verifiable
> ```
</details>
Sorry, not sure I quite follow that :). 

I see that  Install-ADDSDomainController is invoked from MSFT_xADDomainController.psm1.

So if we remove the InModuleScope at the beginning of the try statetement above

     InModuleScope $Global:DSCResourceName {

And then add -ModuleName to the definition of the mock below

    Mock Install-ADDSDomainController -ModuleName MSFT_xADDomainController.psm1 -MockWith {} -ParameterFilter { $SiteName -eq $correctSiteName }

Then the mock will not be called unless we define this fake function?

    function Install-ADDSDomainController { ... }

If that is the case I can try to debug the test. There should be no need to define the fake function.

Thanks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

johlju commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 136 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KORWbzPcHYpOLe6hjwe:-KOfibgKx9yQOWq0R4no:b-cxx9fv) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L136)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> Sorry, not sure I quite follow that :). 
> 
> I see that  Install-ADDSDomainController is invoked from MSFT_xADDomainController.psm1.
> 
> So if we remove the InModuleScope at the beginning of the try statetement above
> 
>      InModuleScope $Global:DSCResourceName {
> 
> And then add -ModuleName to the definition of the mock below
> 
>     Mock Install-ADDSDomainController -ModuleName MSFT_xADDomainController.psm1 -MockWith {} -ParameterFilter { $SiteName -eq $correctSiteName }
> 
> Then the mock will not be called unless we define this fake function?
> 
>     function Install-ADDSDomainController { ... }
> 
> If that is the case I can try to debug the test. There should be no need to define the fake function.
> 
> Thanks
> 
> 
</details>
Exactly, my findings when doing mock's for the xSQLServer resource showed me that a mock will not be called unless you have a fake function that can be invoked. I found this strange also. When I saw your comment I got interested since I had to do the same in my tests as the author of this PR. In this case maybe it works without the fake function, if so, I'm very interested how it's done. Or maybe the module which has the cmdlet is by default on the test machine, which might make it work.
If the command was not faked, I got "command not found" like error.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

narrieta commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 136 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KORWbzPcHYpOLe6hjwe:-KOflHvGnFYl2jTHs_yX:b6plh7n) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L136)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> Exactly, my findings when doing mock's for the xSQLServer resource showed me that a mock will not be called unless you have a fake function that can be invoked. I found this strange also. When I saw your comment I got interested since I had to do the same in my tests as the author of this PR. In this case maybe it works without the fake function, if so, I'm very interested how it's done. Or maybe the module which has the cmdlet is by default on the test machine, which might make it work.
> If the command was not faked, I got "command not found" like error.
</details>
I think in this specific test it should work without the fake.

If you have an instance where it doesn't work I can debugging it to understand what is going on.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

narrieta commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 136 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KORWbzPcHYpOLe6hjwe:-KOflZZnAExFp3AMvnUd:b9wsbcw) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L136)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> I think in this specific test it should work without the fake.
> 
> If you have an instance where it doesn't work I can debugging it to understand what is going on.
</details>
Meant "... I can help debugging ..."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

johlju commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 136 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KORWbzPcHYpOLe6hjwe:-KOfm6GpwzJG8bjdv1Ah:blm9hrg) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L136)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> Meant "... I can help debugging ..."
</details>
Yes, absolutely. Would love feedback. I we can skip the fake function, I wuld be happy. Look at the PR I sent in earlier. https://github.com/PowerShell/xSQLServer/pull/98/files#diff-bd056a9225a6aac31117f6dcd5b22bedR254
We can continue the discussion there.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

narrieta commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 136 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KORWbzPcHYpOLe6hjwe:-KOg3G-u2KN4nIl3YtqA:b-haad5r) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L136)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> Yes, absolutely. Would love feedback. I we can skip the fake function, I wuld be happy. Look at the PR I sent in earlier. https://github.com/PowerShell/xSQLServer/pull/98/files#diff-bd056a9225a6aac31117f6dcd5b22bedR254
> We can continue the discussion there.
</details>
@gordonbondon Please see reply on @johlju 's PR. May be the same case here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

gordonbondon commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 136 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KORWbzPcHYpOLe6hjwe:-KOjmswNncKpFYgp7bFS:b-91nxk4) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L136)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> @gordonbondon Please see reply on @johlju 's PR. May be the same case here.
</details>
The thing is that we actually always have ActiveDirectory module installed on appveyor vm because we install AD-RSAT from windows features.

I've only faked this cmdlet because even if you install AD-RSAT, `Install-ADDomainController` will not be available on client systems like Win 10, Win 8 etc so test will fail when you are developing locally. It is only available on Server 2012 i think.

There are a couple of server only cmdlets thet are faked in xADDomain resource tests like `Install-ADDSForest`, `Install-ADDSDomain`
There are also some redundant fakes in MSFT_xADOrganizationalUnit.Tests.ps1 that can probably be removed.

So my questions. I think faking entire module is not reasonable because there are a lot of functions that are working fine. Should I just fake those functions available only for server in separate file and dot source it? Or nevertheless should we fake the module to make test independent of RSAT installation?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

iainbrighton commented:
There are AD specific exceptions caught that rely on the RSAT tools being installed. There's no way to fake them, hence why the module is installed prior to running the tests. We could remove catching specific exceptions with a general catch, but I'm not sure that's a good idea..

gordonbondon commented:
*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 29 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KOQyY40dxbcEtFv2XEx:-KOk-jPFCROxDYAAUB96:buobj0l) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L29)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> You would need to add -ModuleName to the definition of the mock. There are some examples here: https://github.com/pester/Pester/wiki/Unit-Testing-within-Modules
</details>
Thanks. This should work. 
In tests for `Test-DSCResource` I need to mock internal non-exported function `Test-ADRepolicationSite`. Should i move all the tests for `Test-DSCResource` in `InModuleScope` region?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

gordonbondon commented:
So we should fake just those functions. I'll add throw to the function. Doese it make sence to create separate file for just a couple of functions?

narrieta commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 29 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KOQyY40dxbcEtFv2XEx:-KOkBPI1XZzI67pGqW2X:bugiigk) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L29)):*
<details><summary><i>Previously, gordonbondon (Artem Iarmoliuk) wrote…</i></summary>
> Thanks. This should work. 
> In tests for `Test-DSCResource` I need to mock internal non-exported function `Test-ADRepolicationSite`. Should i move all the tests for `Test-DSCResource` in `InModuleScope` region?
</details>
The preference is to reduce the code within InModuleScope to the minimum that is required. I would move only the tests for the private Test-ADRepolicationSite.

InModuleScope can be nested within a describe.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

narrieta commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 136 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KORWbzPcHYpOLe6hjwe:-KOkCFeSx9WYjFxo9GWa:bit44xx) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L136)):*
<details><summary><i>Previously, gordonbondon (Artem Iarmoliuk) wrote…</i></summary>
> The thing is that we actually always have ActiveDirectory module installed on appveyor vm because we install AD-RSAT from windows features.
> 
> I've only faked this cmdlet because even if you install AD-RSAT, `Install-ADDomainController` will not be available on client systems like Win 10, Win 8 etc so test will fail when you are developing locally. It is only available on Server 2012 i think.
> 
> There are a couple of server only cmdlets thet are faked in xADDomain resource tests like `Install-ADDSForest`, `Install-ADDSDomain`
> There are also some redundant fakes in MSFT_xADOrganizationalUnit.Tests.ps1 that can probably be removed.
> 
> So my questions. I think faking entire module is not reasonable because there are a lot of functions that are working fine. Should I just fake those functions available only for server in separate file and dot source it? Or nevertheless should we fake the module to make test independent of RSAT installation?
</details>
Either option is fine with me. My concern was more that the intent of the fake function wasn't clear. The fake threw me off, I thought that it was actually a leftover from an iteration before you added the mock. Let's add comments documenting the reason of the fake.

Now, and this is just a suggestion, my preference would be to move the definition of the fake out of the test itself, to improve the readability of the test. If I anticipated more tests using the same fakes I would move them to a separate psm1 file and import the module (rather than a ps1 file and dot-source the script).

As far as faking the entire module: the unit tests should not require external setup (e.g. installing RSAT). They should do the setup/cleanup that is needed to execute the tests, whether it be installing RSAT or using a fake module.

Thanks for the updates!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

narrieta commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 136 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KORWbzPcHYpOLe6hjwe:-KOkFYzPizx4pxKUAs9O:b-edzohy) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L136)):*
>  If I anticipated more tests using the same fakes I would move them to a separate psm1 file and import the module (rather than a ps1 file and dot-source the script).

... Or move them to the top of the file (outside the first Describe) if they would be used only within this script ...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

gordonbondon commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 29 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KOQyY40dxbcEtFv2XEx:-KOkVhIaTourkCQZ5dg9:b-3fwycv) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L29)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> The preference is to reduce the code within InModuleScope to the minimum that is required. I would move only the tests for the private Test-ADRepolicationSite.
> 
> InModuleScope can be nested within a describe.
</details>
Done. It appears there is an issue with -ModuleName scoping pester/Pester#204 that prevent using variables from script in mocks. So I had do define all stubs inside mock scriptblocks.

---

*[Tests/Unit/MSFT_xADDomainController.Tests.ps1, line 136 \[r6\]](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-KORWbzPcHYpOLe6hjwe:-KOr_8ztpw8M86kKSmFc:bq56btk) ([raw file](https://github.com/powershell/xactivedirectory/blob/828397d590ea26e8cf08d855ecf2f04481853f26/Tests/Unit/MSFT_xADDomainController.Tests.ps1#L136)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> >  If I anticipated more tests using the same fakes I would move them to a separate psm1 file and import the module (rather than a ps1 file and dot-source the script).
> 
> ... Or move them to the top of the file (outside the first Describe) if they would be used only within this script ...
</details>
Done.
I don't think `Install-ADDomainController` will be used anywhere else so i've moved it to Init part of test and added throw.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109)*
<!-- Sent from Reviewable.io -->

eshaparmar commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 4 files at r1, 1 of 1 files at r5, 1 of 2 files at r6, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xactivedirectory/109#-:-KSEDrSdqTe2aSu5KN9J:bnfp4nl)*
<!-- Sent from Reviewable.io -->
